src/shares_monitor.rs

```diff
@@ -0,0 +1,151 @@
+    batch_size: u32,
+}
+
+impl SharesMonitor {
+    pub fn new() -> Self {
+        SharesMonitor {
+            pending_shares: Arc::new(Mutex::new(Vec::new())),
+            batch_size: BATCH_SIZE, // Default batch size
+        }
+    }
+
+    /// Inserts a new share into the pending shares list.
+    pub fn insert_share(&self, share: SharesInfo) {
+        self.pending_shares
+            .safe_lock(|shares| {
+                shares.push(share);
+            })
+            .unwrap_or_else(|e| {
+                error!("Failed to lock pending shares: {:?}", e);
+                ProxyState::update_downstream_state(DownstreamType::TranslatorDownstream);
+            });
+    }
+
+    /// Retrieves the list of pending shares.
+    fn get_pending_shares(&self) -> Vec<SharesInfo> {
+        self.pending_shares
+            .safe_lock(|shares| shares.clone())
+            .unwrap_or_else(|e| {
+                error!("Failed to lock pending shares: {:?}", e);
+                ProxyState::update_downstream_state(DownstreamType::TranslatorDownstream);
+                Vec::new()
+            })
+    }
+
+    /// Clears the list of pending shares.
+    fn clear_pending_shares(&self) {
+        self.pending_shares
+            .safe_lock(|shares| {
+                shares.clear();
+            })
+            .unwrap_or_else(|e| {
+                error!("Failed to lock pending shares: {:?}", e);
+                ProxyState::update_downstream_state(DownstreamType::TranslatorDownstream);
+            });
+    }
+
+    /// Monitors the pending shares and sends them to the monitoring server in batches.
+    pub async fn monitor(&self) {
+        let api = MonitorAPI::new(MONITORING_SERVER_URL);
+        let mut interval = tokio::time::interval(std::time::Duration::from_secs(60)); // Check every 60 seconds
+        loop {
+            interval.tick().await;
+            let shares_to_send = self.get_pending_shares();
+            if !shares_to_send.is_empty() {
+                if shares_to_send.len() >= self.batch_size as usize {
+                    debug!("Sending batch of {} shares to API", shares_to_send.len());
+                    match api.send_shares(shares_to_send.clone()).await {
+                        Ok(_) => {
+                            debug!("Successfully sent Shares: {:?} to API", &shares_to_send);
+                            self.clear_pending_shares(); // Clear after sending
+                        }
+                        Err(e) => {
+                            error!("{}", Error::ApiError(e.to_string()));
+                        }
+                    }
+                } else {
+                    debug!("Not enough shares to send, waiting for more");
+                }
+            } else {
+                debug!("No pending shares to send");
+            }
+        }
+    }
+}
+
+struct MonitorAPI {
+    url: String,
+    client: reqwest::Client,
+}
+
+impl MonitorAPI {
+    fn new(url: &str) -> Self {
```

Comment:
just read the url from 'env' from withing the new function, no need it for it to be an input

Reply:
On the other-hand, I do understand why u wanna go with a config arg. I am not totally against it to be honest. Adding env vars to this project might just make it too complex.